Deleting a config repo causes PipelineNotFoundException in the background

##### Issue Type
- Bug Report

##### Summary
1. Create a config repo from the SPA
2. Wait for it to get parsed and show up on the dashboard
3. Delete the config repo from the SPA
4. Check Go server logs. For the deleted pipelines, we repeatedly see:
```
java.lang.RuntimeException: com.thoughtworks.go.config.PipelineNotFoundException: Pipeline 'yaml-example' not found.
	at com.thoughtworks.go.server.scheduling.ScheduleCheckListener.onMessage(ScheduleCheckListener.java:48)
	at com.thoughtworks.go.server.scheduling.ScheduleCheckListener.onMessage(ScheduleCheckListener.java:25)
	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:73)
	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:53)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: com.thoughtworks.go.config.PipelineNotFoundException: Pipeline 'yaml-example' not found.
	at com.thoughtworks.go.config.BasicCruiseConfig.pipelineConfigByName(BasicCruiseConfig.java:615)
	at com.thoughtworks.go.server.service.GoConfigService.pipelineConfigNamed(GoConfigService.java:252)
	at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.autoSchedulePipeline(BuildCauseProducerService.java:110)
	at com.thoughtworks.go.server.scheduling.ScheduleCheckListener.onMessage(ScheduleCheckListener.java:44)
	... 4 common frames omitted
```

--- a/server/src/main/java/com/thoughtworks/go/server/service/PipelineScheduler.java
+++ b/server/src/main/java/com/thoughtworks/go/server/service/PipelineScheduler.java
@@ -17,6 +17,7 @@
 package com.thoughtworks.go.server.service;
 
 import com.thoughtworks.go.config.*;
+import com.thoughtworks.go.config.remote.ConfigRepoConfig;
 import com.thoughtworks.go.domain.PiplineConfigVisitor;
 import com.thoughtworks.go.listener.ConfigChangedListener;
 import com.thoughtworks.go.listener.EntityConfigChangedListener;
@@ -40,6 +41,7 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Consumer;
 
 @Service
 public class PipelineScheduler implements ConfigChangedListener, GoMessageListener<ScheduleCheckCompletedMessage> {
@@ -77,6 +79,7 @@ public class PipelineScheduler implements ConfigChangedListener, GoMessageListen
     public void initialize() {
         goConfigService.register(this);
         goConfigService.register(pipelineConfigChangedListener());
+        goConfigService.register(configRepoConfigChangedListener());
         scheduleCheckCompletedTopic.addListener(this);
     }
 
@@ -90,6 +93,16 @@ public class PipelineScheduler implements ConfigChangedListener, GoMessageListen
                         pipelines.remove(pipelineConfig.name().toString());
                     }
                 }
+            }
+        };
+    }
+
+    protected EntityConfigChangedListener<ConfigRepoConfig> configRepoConfigChangedListener() {
+        final Consumer<CruiseConfig> configChangeConsumer = this::onConfigChange;
+        return new EntityConfigChangedListener<ConfigRepoConfig>() {
+            @Override
+            public void onEntityConfigChange(ConfigRepoConfig configRepoConfig) {
+                configChangeConsumer.accept(goConfigService.cruiseConfig());
             }
         };
     }

--- a/server/src/test-fast/java/com/thoughtworks/go/server/service/PipelineSchedulerTest.java
+++ b/server/src/test-fast/java/com/thoughtworks/go/server/service/PipelineSchedulerTest.java
@@ -16,9 +16,16 @@
 
 package com.thoughtworks.go.server.service;
 
+import com.thoughtworks.go.config.BasicCruiseConfig;
 import com.thoughtworks.go.config.CaseInsensitiveString;
 import com.thoughtworks.go.config.PipelineConfig;
+import com.thoughtworks.go.config.remote.ConfigRepoConfig;
+import com.thoughtworks.go.config.remote.ConfigReposConfig;
+import com.thoughtworks.go.config.remote.PartialConfig;
+import com.thoughtworks.go.config.remote.RepoConfigOrigin;
 import com.thoughtworks.go.domain.materials.MaterialConfig;
+import com.thoughtworks.go.helper.MaterialConfigsMother;
+import com.thoughtworks.go.helper.PartialConfigMother;
 import com.thoughtworks.go.helper.ScheduleCheckMessageMatcher;
 import com.thoughtworks.go.listener.ConfigChangedListener;
 import com.thoughtworks.go.listener.EntityConfigChangedListener;
@@ -38,6 +45,7 @@ import org.mockito.internal.verification.NoMoreInteractions;
 import java.util.*;
 
 import static com.thoughtworks.go.helper.GoConfigMother.configWithPipelines;
+import static java.util.Arrays.asList;
 import static org.hamcrest.core.Is.is;
 import static org.junit.Assert.assertThat;
 import static org.mockito.Mockito.*;
@@ -198,5 +206,47 @@ public class PipelineSchedulerTest {
         entityConfigChangedListener.onEntityConfigChange(newPipeline);
         scheduler.checkPipelines();
         verify(queue, times(0)).post(ScheduleCheckMessageMatcher.matchScheduleCheckMessage(pipelineName));
+    }
+
+    @Test
+    public void shouldRemovePipelinesOnConfigRepoDeletion() {
+        //setup 2 config repos
+        BasicCruiseConfig cruiseConfig = new BasicCruiseConfig();
+        ConfigRepoConfig repoConfig1 = new ConfigRepoConfig(MaterialConfigsMother.gitMaterialConfig("url1"), "plugin");
+        ConfigRepoConfig repoConfig2 = new ConfigRepoConfig(MaterialConfigsMother.gitMaterialConfig("url2"), "plugin");
+        cruiseConfig.setConfigRepos(new ConfigReposConfig(repoConfig1, repoConfig2));
+        PartialConfig partialConfigInRepo1 = PartialConfigMother.withPipeline("pipeline_in_repo1", new RepoConfigOrigin(repoConfig1, "repo1_r1"));
+        PartialConfig partialConfigInRepo2 = PartialConfigMother.withPipeline("pipeline_in_repo2", new RepoConfigOrigin(repoConfig2, "repo2_r1"));
+        cruiseConfig.merge(asList(partialConfigInRepo1, partialConfigInRepo2), false);
+
+        when(configService.cruiseConfig()).thenReturn(cruiseConfig);
+        ArgumentCaptor<ConfigChangedListener> captor = ArgumentCaptor.forClass(ConfigChangedListener.class);
+        doNothing().when(configService).register(captor.capture());
+        scheduler.initialize();
+        List<ConfigChangedListener> listeners = captor.getAllValues();
+        listeners.get(0).onConfigChange(cruiseConfig);
+        assertThat(listeners.contains(scheduler), is(true));
+        assertThat(listeners.get(2) instanceof EntityConfigChangedListener, is(true));
+        EntityConfigChangedListener<ConfigRepoConfig> entityConfigChangedListener = (EntityConfigChangedListener<ConfigRepoConfig>) listeners.get(2);
+
+        //both should get scheduled
+        scheduler.checkPipelines();
+        verify(queue, times(2)).post(any(ScheduleCheckMessage.class));
+
+        //reset
+        scheduler.onMessage(new ScheduleCheckCompletedMessage("pipeline_in_repo1", 0));
+        scheduler.onMessage(new ScheduleCheckCompletedMessage("pipeline_in_repo2", 0));
+        reset(queue);
+
+        //remove one config repo
+        BasicCruiseConfig updatedConfig = new BasicCruiseConfig();
+        updatedConfig.setConfigRepos(new ConfigReposConfig(repoConfig1, repoConfig2));
+        updatedConfig.merge(Collections.singletonList(partialConfigInRepo1), false);
+        when(configService.cruiseConfig()).thenReturn(updatedConfig);
+        entityConfigChangedListener.onEntityConfigChange(null);
+
+        //removed pipeline should not get scheduled
+        scheduler.checkPipelines();
+        verify(queue, times(1)).post(any(ScheduleCheckMessage.class));
     }
 }

